compiler/rustc_target/src/spec/aarch64_unknown_uefi.rs

```diff
@@ -0,0 +1,36 @@
+// This defines the amd64 target for UEFI systems as described in the UEFI specification. See the
+// uefi-base module for generic UEFI options. On aarch64 systems UEFI systems always run in long-mode,
+// have the interrupt-controller pre-configured and force a single-CPU execution.
+//
+// The win64 ABI is used. It differs from the sysv64 ABI, so we must use a windows target with
+// LLVM. "aarch64-unknown-windows" is used to get the minimal subset of windows-specific features.
+
+use super::uefi_msvc_base;
+use crate::spec::{CodeModel, LinkerFlavor, LldFlavor, Target};
+
+pub fn target() -> Target {
+    let mut base = uefi_msvc_base::opts();
+
+    base.max_atomic_width = Some(64);
+
+    let pre_link_args_msvc = vec!["/machine:arm64".to_string()];
+
+    base.pre_link_args.get_mut(&LinkerFlavor::Msvc).unwrap().extend(pre_link_args_msvc.clone());
+    base.pre_link_args
+        .get_mut(&LinkerFlavor::Lld(LldFlavor::Link))
+        .unwrap()
+        .extend(pre_link_args_msvc);
+
+    // UEFI systems run without a host OS, hence we cannot assume any code locality. We must tell
+    // LLVM to expect code to reference any address in the address-space. The "large" code-model
+    // places no locality-restrictions, so it fits well here.
+    base.code_model = Some(CodeModel::Large);
```

Comment:
I suspect that this is a misunderstanding (affecting all UEFI targets).
Applications running under an OS also cannot assume code locality, they can span the whole virtual memory if necessary.

Applications don't usually take gigabytes of memory though, so the defaults are different and removing this should result in better codegen (comparable with majority of other targets).
The only problem is that the reasoning above applies to ELF targets, and I'm not actually sure whether this option does anything at all if the target is non-ELF.
cc @dvdhrm

Reply:
UEFI applications run in a single address-space, everything is identity-mapped. You might get any pointer value at runtime and access it, or jump to it. You cannot tell up front how your address-space will look like. And since the entire UEFI core is running in your address-space, you cannot even tell what kind of instructions will be invoked as part of your application.

Anyway, the biggest issue I had was that I had to select something for the code-model field, and I found it quite hard to get information on what it controls. Does this only affect immediate-values of generated instructions? Or only the hard-coded relative offsets? Does it affect program-behavior? If it only affects generated constants, why can't the compiler figure it out automatically? What happens when the value is wrongly chosen?

Given all those questions, I figured that `large` was the safest compromise. If you know LLVM (or x86 ISA) better than that, please feel free to help us out here!

Reply:
I am pretty sure we require the `large` code model as the comments above say and as @dvdhrm said it is the safest compromise.

Reply:
Ok.

Reply:
> Anyway, the biggest issue I had was that I had to select something for the code-model field, and I found it quite hard to get information on what it controls.

AFAIK it basically controls when to use relative and when to use absolute addressing. And if relative how much bits is should leave for the linker to fill in the offset.